server/api/animelist.ts

```diff
@@ -0,0 +1,174 @@
+import axios from 'axios';
+import xml2js from 'xml2js';
+import fs, { promises as fsp } from 'fs';
+import path from 'path';
+import logger from '../logger';
+
+const UPDATE_INTERVAL = 24 * 3600; // how often to download new mapping
+// originally at https://raw.githubusercontent.com/ScudLee/anime-lists/master/anime-list.xml
+const MAPPING_URL =
+  'https://raw.githubusercontent.com/Anime-Lists/anime-lists/master/anime-list.xml';
+const LOCAL_PATH = path.join(__dirname, '../../config/anime-list.xml');
+
+const mappingRegexp = new RegExp(/;[0-9]+-([0-9]+);/);
+
+// Anime-List xml files are community maintained mappings that Hama agent uses to map AniDB IDs to tvdb/tmdb IDs
+// https://github.com/Anime-Lists/anime-lists/
+
+export interface AnidbItem {
+  tvdbId?: number;
+  tmdbId?: number;
+  imdbId?: string;
+}
+
+class AnimeListMapping {
+  private syncing = false;
+
+  private mapping: { [anidbId: number]: AnidbItem } = {};
+
+  // each episode in season 0 from TVDB can map to movie (AniDB id)
+  private specials: { [tvdbId: number]: { [episode: number]: number } } = {};
+
+  public isLoaded = () => Object.keys(this.mapping).length !== 0;
+
+  private loadFromFile = async () => {
+    logger.info('Loading Anime-List mapping file');
+    try {
+      const file = await fsp.readFile(LOCAL_PATH);
+      const xml = await xml2js.parseStringPromise(file);
```

Comment:
Can we add some types for what we expect this XML to look like? We don't have to type everything exactly, just what we are using. Should be less error prone to things like typos in the future.

Reply:
I was kind of afraid you'll ask for this... I tried to do that initially, but I failed terrible due to not knowing typescript so I gave up. Let me try again.

Reply:
Bug me in discord if you need any help with it!